docs/design/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,269 @@
+
+	// current number of replicas of pods managed by this autoscaler.
+	CurrentReplicas int
+
+	// desired number of replicas of pods managed by this autoscaler.
+	DesiredReplicas int
+
+	// current average CPU utilization over all pods, represented as a percentage of requested CPU,
+	// e.g. 70 means that an average pod is using now 70% of its requested CPU.
+	CurrentCPUUtilizationPercentage *int
+}
+```
+
+``Scale`` is a reference to the Scale subresource.
+``MinReplicas``, ``MaxReplicas`` and ``CPUUtilization`` define autoscaler configuration.
+We are also introducing HorizontalPodAutoscalerList object to enable listing all autoscalers in the cluster:
+
+```go
+// list of horizontal pod autoscaler objects.
+type HorizontalPodAutoscalerList struct {
+	unversioned.TypeMeta
+	unversioned.ListMeta
+
+	// list of horizontal pod autoscaler objects.
+	Items []HorizontalPodAutoscaler
+}
+```
+
+## Autoscaling Algorithm
+
+The autoscaler is implemented as a control loop. It periodically queries pods described by ```Status.PodSelector``` of Scale subresource, and collects their CPU utilization.
+Then, it compares the arithmetic mean of the pods' CPU utilization with the target defined in ``Spec.CPUUtilization``,
+and adjust the replicas of the Scale if needed to match the target
+(preserving condition: MinReplicas <= Replicas <= MaxReplicas).
+
+The period of the autoscaler is controlled by ``--horizontal-pod-autoscaler-sync-period`` flag of controller manager.
+The default value is 30 seconds.
+
+
+CPU utilization is the recent CPU usage of a pod (average across the last 1 minute) divided by the CPU requested by the pod.
+In Kubernetes version 1.1, CPU usage is taken directly from Heapster.
+In future, there will be API on master for this purpose
+(see [#11951](https://github.com/kubernetes/kubernetes/issues/11951)).
+
+The target number of pods is calculated from the following formula:
+
+```
+TargetNumOfPods = ceil(sum(CurrentPodsCPUUtilization) / Target)
+```
+
+Starting and stopping pods may introduce noise to the metric (for instance, starting may temporarily increase CPU).
+So, after each action, the autoscaler should wait some time for reliable data.
+Scale-up will happen if there was no rescaling within the last 3 minutes.
+Scale-down will wait for 5 minutes from the last rescaling. Moreover any scaling will only be made if: ``avg(CurrentPodsConsumption) / Target`` drops below 0.9 or increases above 1.1 (10% tolerance). Such approach has two benefits:
+
+* Autoscaler works in a conservative way.
+  If new user load appears, it is important for us to rapidly increase the number of pods,
+  so that user requests will not be rejected.
+  Lowering the number of pods is not that urgent.
+
+* Autoscaler avoids thrashing, i.e.: prevents rapid execution of conflicting decision if the load is not stable.
+
+## Relative vs. absolute metrics
+
+The question arises whether the values of the target metric should be absolute (e.g.: 0.6 core)
```

Comment:
Replace first sentence with:
We chose values of the target metric to be relative (e.g. 90% of requested CPU resource) rather than absolute (e.g. 0.6 core) for the following reason.